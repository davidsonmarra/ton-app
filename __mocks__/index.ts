export { productDetailsApi } from './product-details-api';
export { productsApi } from './products-api';
