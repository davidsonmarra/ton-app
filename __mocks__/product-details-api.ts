export const productDetailsApi = [
  {
    id: 1,
    title: 'T1',
    subTitle: 'Promo',
    oldPrice: '198,80',
    currentPrice: '106,80',
    installmentsValue: '8',
    installmentsValueCents: '90',
    images: [
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t1-1',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t1-2',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t1-3'
    ],
    benefits: [
      {
        icon: 'calendar-blank-outline',
        description: 'Receba suas vendas em 1 dia útil'
      },
      {
        icon: 'truck-outline',
        description: 'Frete grátis para todo o Brasil'
      },
      {
        icon: 'cellphone-check',
        description: 'Precisa de celular com bluetooth e conectado'
      },

      {
        icon: 'chat-alert-outline',
        description: 'Comprovante via aplicativo e SMS'
      }
    ]
  },
  {
    id: 2,
    title: 'T1 Chip',
    subTitle: 'Promo',
    oldPrice: '318,80',
    currentPrice: '286,80',
    installmentsValue: '23',
    installmentsValueCents: '90',
    images: [
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t1-chip-1',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t1-chip-2',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t1-chip-3'
    ],
    benefits: [
      {
        icon: 'calendar-blank-outline',
        description: 'Receba suas vendas em 1 dia útil'
      },
      {
        icon: 'truck-outline',
        description: 'Frete grátis para todo o Brasil'
      },
      {
        icon: 'cellphone-remove',
        description: 'Não precisa de um celular'
      },
      {
        icon: 'chat-alert-outline',
        description: 'Comprovante via aplicativo e SMS'
      },
      {
        icon: 'wifi',
        description: 'Conexão Wi-Fi e chip de dados'
      },
      {
        icon: 'nfc',
        description: 'Pagamento por aproximação (NFC)'
      }
    ]
  },
  {
    id: 3,
    title: 'T2+',
    subTitle: 'Promo',
    oldPrice: '478,80',
    currentPrice: '190,80',
    installmentsValue: '15',
    installmentsValueCents: '90',
    images: [
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t2-1',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t2-2',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t2-3'
    ],
    benefits: [
      {
        icon: 'calendar-blank-outline',
        description: 'Receba suas vendas em 1 dia útil'
      },
      {
        icon: 'truck-outline',
        description: 'Frete grátis para todo o Brasil'
      },
      {
        icon: 'cellphone-remove',
        description: 'Não precisa de um celular'
      },
      {
        icon: 'chat-alert-outline',
        description: 'Comprovante via aplicativo e SMS'
      },
      {
        icon: 'wifi',
        description: 'Conexão Wi-Fi e chip de dados'
      },
      {
        icon: 'nfc',
        description: 'Pagamento por aproximação (NFC)'
      },
      {
        icon: 'battery-high',
        description: 'Bateria de longa duração (+ de 12h)'
      }
    ]
  },
  {
    id: 4,
    title: 'T3',
    subTitle: 'Promo',
    oldPrice: '958,80',
    currentPrice: '394,80',
    installmentsValue: '32',
    installmentsValueCents: '90',
    images: [
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t3-1',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t3-2',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_auto/site-ton/maquininhas/machine-t3-3'
    ],
    benefits: [
      {
        icon: 'calendar-blank-outline',
        description: 'Receba suas vendas em 1 dia útil'
      },
      {
        icon: 'truck-outline',
        description: 'Frete grátis para todo o Brasil'
      },
      {
        icon: 'cellphone-remove',
        description: 'Não precisa de um celular'
      },
      {
        icon: 'newspaper',
        description: 'Comprovante impresso ou SMS'
      },
      {
        icon: 'wifi',
        description: 'Conexão Wi-Fi e chip de dados'
      },
      {
        icon: 'nfc',
        description: 'Pagamento por aproximação (NFC)'
      },
      {
        icon: 'qrcode',
        description: 'Aceita Pix QR Code na Maquininha'
      }
    ]
  },
  {
    id: 5,
    title: 'T3 Smart',
    subTitle: 'Promo',
    oldPrice: '958,80',
    currentPrice: '478,80',
    installmentsValue: '39',
    installmentsValueCents: '90',
    images: [
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_100/site-ton/maquininhas/release-machine-t3-smart-1',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_100/site-ton/maquininhas/release-machine-t3-smart-2',
      'https://res.cloudinary.com/dunz5zfpt/fl_progressive/f_auto,c_limit,w_384,q_100/site-ton/maquininhas/release-machine-t3-smart-3'
    ],
    benefits: [
      {
        icon: 'calendar-blank-outline',
        description: 'Receba suas vendas em 1 dia útil'
      },
      {
        icon: 'truck-outline',
        description: 'Frete grátis para todo o Brasil'
      },
      {
        icon: 'qrcode',
        description: 'Aceita Pix QR Code na Maquininha'
      },
      {
        icon: 'cellphone-remove',
        description: 'Não precisa de um celular'
      },
      {
        icon: 'newspaper',
        description: 'Comprovante impresso ou SMS'
      },
      {
        icon: 'nfc',
        description: 'Pagamento por aproximação (NFC)'
      },
      {
        icon: 'wifi',
        description: 'Conexão Wi-Fi e chip de dados'
      },
      {
        icon: 'battery-high',
        description: 'Bateria de longa duração'
      },
      {
        icon: 'android',
        description: 'Sistema Android com Visor Touchscreen'
      }
    ]
  }
];
