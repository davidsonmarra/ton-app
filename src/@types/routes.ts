export type ProductsStackParamList = {
  HomeScreen: undefined;
  ProductDetailsScreen: { id: number };
};
