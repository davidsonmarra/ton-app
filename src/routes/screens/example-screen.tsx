import React from 'react';
import { Example } from '../../presentational';

export const ExampleScreen = () => <Example />;
