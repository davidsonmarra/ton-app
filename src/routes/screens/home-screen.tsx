import React from 'react';
import { Home } from '../../presentational';

export const HomeScreen = () => <Home />;
