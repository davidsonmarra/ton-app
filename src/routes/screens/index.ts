export { ExampleScreen } from './example-screen';
export { ProductDetailsScreen } from './product-details-screen';
export { ProductsListScreen } from './products-list-screen';
