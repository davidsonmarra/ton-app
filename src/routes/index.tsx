import React from 'react';
import { NavigationContainer } from '@react-navigation/native';
import { ProductsRoutes } from './stacks/products.routes';

export function Routes() {
  return (
    <NavigationContainer>
      <ProductsRoutes />
    </NavigationContainer>
  );
}
