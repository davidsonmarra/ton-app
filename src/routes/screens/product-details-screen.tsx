import React, { useCallback, useLayoutEffect, useState } from 'react';
import { ProductDetailsDTO, ProductsStackParamList } from '../../@types';
import { ProductDetails } from '../../presentational';
import { NavigationProp, RouteProp, useNavigation, useRoute } from '@react-navigation/native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { api } from '../../global/services';

export const ProductDetailsScreen = () => {
  const [product, setProduct] = useState({} as ProductDetailsDTO);
  const [isLoading, setIsLoading] = useState(true);

  const { bottom } = useSafeAreaInsets();
  const { goBack } =
    useNavigation<NavigationProp<ProductsStackParamList, 'ProductDetailsScreen'>>();
  const { id } = useRoute<RouteProp<ProductsStackParamList, 'ProductDetailsScreen'>>().params;

  const getProduct = useCallback(async () => {
    try {
      const { data } = await api.get(`products/${id}`);
      setProduct(data);
    } catch (error) {
      console.log(error);
      goBack();
    }
    setIsLoading(false);
  }, [product]);

  const handlePressLeftIcon = useCallback(() => {
    goBack();
  }, []);

  useLayoutEffect(() => {
    getProduct();
  }, []);

  return (
    <ProductDetails
      isLoadingApi={isLoading}
      item={product}
      handlePressLeftIcon={handlePressLeftIcon}
      bottomInset={bottom}
    />
  );
};
