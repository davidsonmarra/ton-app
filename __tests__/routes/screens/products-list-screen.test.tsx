import { productsApi } from '../../../__mocks__';
import { ProductsListScreen } from '../../../src/routes/screens';
import { matchSnapshotWithProviders, render, act, waitFor, fireEvent } from '../../test-utils';

const mockFetch = jest.fn().mockResolvedValue({
  data: { products: productsApi }
});

jest.mock('../../../src/global/services', () => ({
  api: {
    get: async () => mockFetch()
  }
}));

const mockNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({
    navigate: mockNavigate,
    goBack: jest.fn()
  }),
  useRoute: () => ({
    params: {
      id: 1
    }
  })
}));

describe('ProductListScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match a snapshot', async () => {
    await act(() => {
      matchSnapshotWithProviders(<ProductsListScreen />);
    });
  });

  it('should render list correctly', async () => {
    await waitFor(async () => {
      const { findByText } = render(<ProductsListScreen />);
      expect(await findByText('Maquininha ' + productsApi[0].name)).toBeTruthy();
      expect(await findByText('Maquininha ' + productsApi[1].name)).toBeTruthy();
      expect(await findByText('Maquininha ' + productsApi[2].name)).toBeTruthy();
      expect(await findByText('Maquininha ' + productsApi[3].name)).toBeTruthy();
      expect(await findByText('Maquininha ' + productsApi[4].name)).toBeTruthy();
    });
  });

  it('should navigate to ProductDetailsScreen when click on product', async () => {
    await waitFor(async () => {
      const { findByText } = render(<ProductsListScreen />);
      const product = await findByText('Maquininha ' + productsApi[0].name);
      expect(mockNavigate).not.toHaveBeenCalled();
      fireEvent.press(product);
      expect(mockNavigate).toBeCalled();
    });
  });

  it('should call function on api error', async () => {
    mockFetch.mockRejectedValueOnce({ message: 'error' });
    const consoleSpy = jest.spyOn(console, 'log');
    expect(consoleSpy).not.toHaveBeenCalled();
    await waitFor(async () => {
      render(<ProductsListScreen />);
      await waitFor(() => expect(mockFetch).toHaveBeenCalledTimes(1));
      expect(consoleSpy).toHaveBeenCalledTimes(1);
    });
  });
});
