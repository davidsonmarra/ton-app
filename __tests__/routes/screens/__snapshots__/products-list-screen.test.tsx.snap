// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListScreen should match a snapshot 1`] = `null`;
